native/src/modules/fetcher/FetcherModule.js

```diff
@@ -1,47 +1,53 @@
 // @flow
 
-import NativeFetcherModule, { NativeFetcherModuleEmitter } from './NativeFetcherModule'
 import { isEmpty } from 'lodash'
+import type { EventChannel } from 'redux-saga'
+import { eventChannel } from 'redux-saga'
+import NativeFetcherModule, { NativeFetcherModuleEmitter } from './NativeFetcherModule'
 
 export type TargetFilePathsType = { [path: string]: string }
 
 export type FetchResultType = { [path: string]: {| lastUpdate: string, url: string, errorMessage: ?string |} }
 
-export type ProgressCallbackType = (progress: number) => void
-
 class FetcherModule {
-  // TODO NATIVE-264: Correctly handle already fetching
+  // TODO IGAPP-217: Correctly handle already fetching
   static currentlyFetching = false
 
-  async fetchAsync (
-    targetFilePaths: TargetFilePathsType,
-    progress: ProgressCallbackType
-  ): Promise<FetchResultType> {
-    if (FetcherModule.currentlyFetching) {
-      throw new Error('Already fetching!')
-    }
-    if (isEmpty(targetFilePaths)) {
-      return {}
-    }
+  createProgressChannel = (): EventChannel<number> => {
+    return eventChannel<number>(emitter => {
+      let prevStep = 0
+      const stepWidth = 10
+
+      const subscription = NativeFetcherModuleEmitter.addListener('progress', (progress: number) => {
+        const newStep = Math.floor(progress * stepWidth) / stepWidth
+        if (newStep <= prevStep) {
+          return
+        }
+        prevStep = newStep
+        emitter(newStep)
+      })
+
+      return () => subscription.remove()
+    })
+  }
+
+  fetchAsync = (targetFilePaths: TargetFilePathsType): Promise<FetchResultType> => {
```

Comment:
This is not really async anymore, could we revert this function to some extend (without the listener of course)?

Reply:
We can rename it to fetch (without async), depends how we define "async". Inside of the method itself no await is used, so the "async" tag is not needed, that is true. But the result still as async.
The method returns no value itself but a promise, so it behaves in the view of the caller in the same way as a method which awaits.

Did I understand your question correctly? If not perhaps you could give me a suggestion / proposal how to "refactor without listener".

Reply:
You can just await the result here. E.g.:
```js
  fetchAsync = async (targetFilePaths: TargetFilePathsType): Promise<FetchResultType> => {
    if (isEmpty(targetFilePaths)) {
      return {}
    }
    FetcherModule.currentlyFetching = true

    const result = await NativeFetcherModule.fetchAsync(targetFilePaths)

    FetcherModule.currentlyFetching = false

    if (!result) {
      // While testing the app I noticed that the FetchResultType of this function was empty. I am absolutely not sure
      // why this happened. As this cause an inconsistent state it is better to throw an error in this case.
      throw new Error('Fetch failed for some reason!')
    }
    return result
  }

```

Reply:
The proposal by klinzo is equivalent, but more readable. It uses async/await syntax instead of then/catch

Reply:
Yes, looks much cleaner, thanks!